### PR TITLE
Support min max object size in s3_lifecycle module

### LINKED ISSUE #861: s3_lifecycle - Add size thresholds
### Summary

STANDARD_IA and GLACIER classes in S3 include an overhead for access history data. For small files the overhead can be larger than the file itself, so you end up paying more to store small files in these supposedly cheaper classes.

The solution is to add a minimum size filter to the lifecycle config. The API supports this. The s3_lifecycle ansible module does not yet support these. We should add a min size filter (and also max size filter).

### Issue Type

Feature Idea

### Component Name

s3_lifecycle

### Additional Information

In the boto3 call, `ObjectSizeGreaterThan` and `ObjectSizeLessThan` are optional fields which go next to Prefix under Filter. They are the number of bytes (not kB).

I propose this usage:

```
- name: Configure a lifecycle rule on a bucket to expire (delete) items with a prefix of /logs/ after 30 days
  community.aws.s3_lifecycle:
    name: mybucket
    expiration_days: 30
    prefix: logs/
    object_size_greater_than: 123
    object_size_less_than: 123
    status: enabled
    state: present
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by leo4ever**
##### SUMMARY

Support the S3 lifecycle settings of minimum and maximum object size to apply the lifecycle rules.

Fixes #861

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

s3_lifecycle

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

**Review by tremble — CHANGES_REQUESTED** [review]
Thanks for taking the time to submit this, especially for including the new integration tests.

If you take a look at the output from the tests you'll see a couple of issues:

- https://github.com/ansible-collections/community.aws/actions/runs/12383415807job/34597832812?pr=2205
- https://github.com/ansible-collections/community.aws/actions/runs/12383415807/job/34597834846?pr=2205

These boil down to

- The typo on line 1 breaks the documentation (YAML)
- We use the "black" formatting style, If you've got tox setup, then `tox -m format`  it should fix the formatting.

WRT Unit tests:
Unit tests can be a PITA to implement (they live under tests/unit/plugins/modules/).
I'd really like to see some basic tests for filters_are_equal if possible.  However, by moving filters_are_equal outside of compare_and_update_configuration you'll make it easier for someone (possibly me) to come along later and add them if you're not able to do so.

File: plugins/modules/s3_lifecycle.py [hunk 1]
```diff
@@ -59,6 +59,16 @@
         and replaced with the new transition(s)
     default: true
     type: bool
+  maximum_object_size:
+    descriptionL
```
[the remark this answers is not in the document]
Comment (leo4ever) [reply plugins/modules/s3_lifecycle.py]: @tremble - I have made the changes based on your comments. But still seeing the build on collections docs failing - https://github.com/ansible-collections/community.aws/actions/runs/12403295884

I have created a unit test but having trouble trying to run tests locally. Can this test be executed as part of CI?

File: plugins/modules/s3_lifecycle.py [hunk 2]
```diff
@@ -598,6 +629,8 @@ def main():
         expiration_days=dict(type="int"),
         expiration_date=dict(),
         expire_object_delete_marker=dict(type="bool"),
+        maximum_object_size=dict(type=int),
+        minimum_object_size=dict(type=int),
```
Comment (tremble) [inline plugins/modules/s3_lifecycle.py:633]:
```suggestion
        maximum_object_size=dict(type="int"),
        minimum_object_size=dict(type="int"),
```

**Comment by tremble** [comment]
One other minor request:  Please add a "minor_changes" changelog entry: https://docs.ansible.com/ansible/latest/community/development_process.html#changelogs-how-to

File: changelogs/fragments/2205-support-minmax-s3lifecycle.yml [hunk 3]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - s3_lifecycle - Support for min and max object size when applying the filter rules.
```
Comment (tremble) [inline changelogs/fragments/2205-support-minmax-s3lifecycle.yml:2]:
```suggestion
  - s3_lifecycle - Support for min and max object size when applying the filter rules (https://github.com/ansible-collections/community.aws/pull/2205).
```

File: plugins/modules/s3_lifecycle.py [hunk 4]
```diff
@@ -59,6 +59,16 @@
         and replaced with the new transition(s)
     default: true
     type: bool
+  maximum_object_size:
+    description:
+      - The maximum object size to which the rule applies.
+    required: false
+    type: int
```
Comment (tremble) [inline plugins/modules/s3_lifecycle.py:66]:
```suggestion
    type: int
    version_added: 9.1.0
```

File: plugins/modules/s3_lifecycle.py [hunk 5]
```diff
@@ -59,6 +59,16 @@
...
+  minimum_object_size:
+    description:
+      - The minimum object size to which the rule applies.
+    required: false
```
Comment (tremble) [inline plugins/modules/s3_lifecycle.py:70]:
```suggestion
    required: false
    version_added: 9.1.0
```

File: plugins/modules/s3_lifecycle.py [hunk 6]
```diff
@@ -271,13 +281,30 @@ def fetch_rules(client, module, name):
         module.fail_json_aws(e)
     return current_lifecycle_rules
 
-
+# Helper function to deeply compare filters
```
Comment (tremble) [inline plugins/modules/s3_lifecycle.py:284]:
```suggestion

# Helper function to deeply compare filters
```

File: plugins/modules/s3_lifecycle.py [hunk 7]
```diff
@@ -271,13 +281,30 @@ def fetch_rules(client, module, name):
...
+    # Treat empty string as equal to a filter not being set
+    return (
+        filter1.get("Prefix", "") == filter2.get("Prefix", "")
+        and filter1.get("ObjectSizeGreaterThan") == filter2.get("ObjectSizeGreaterThan")
+        and filter1.get("ObjectSizeLessThan") == filter2.get("ObjectSizeLessThan")
+        and filter1.get("And", {}).get("Prefix", "") == filter2.get("And", {}).get("Prefix", "")
+        and filter1.get("And", {}).get("ObjectSizeGreaterThan") == filter2.get("And", {}).get("ObjectSizeGreaterThan")
+        and filter1.get("And", {}).get("ObjectSizeLessThan") == filter2.get("And", {}).get("ObjectSizeLessThan")
+    )
+
```
Comment (tremble) [inline plugins/modules/s3_lifecycle.py:299]:
```suggestion


```

File: plugins/modules/s3_lifecycle.py [hunk 8]
```diff
@@ -362,18 +397,14 @@ def compare_and_update_configuration(client, module, current_lifecycle_rules, ru
     lifecycle_configuration = dict(Rules=[])
     changed = False
     appended = False
-
+
```
Comment (tremble) [inline plugins/modules/s3_lifecycle.py:400]:
```suggestion

```

File: plugins/modules/s3_lifecycle.py [hunk 9]
```diff
@@ -681,4 +714,4 @@ def main():
 
 
 if __name__ == "__main__":
-    main()
+    main()
```
Comment (tremble) [inline plugins/modules/s3_lifecycle.py:717]:
```suggestion
    main()
```

File: tests/unit/plugins/modules/test_s3_lifecycle_filters_are_equal.py [hunk 10]
```diff
@@ -0,0 +1,108 @@
+# -*- coding: utf-8 -*-
+
+# Copyright: (c) 2023, Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+import pytest
+
+from ansible_collections.community.aws.plugins.modules.s3_lifecycle import filters_are_equal
+
```
Comment (tremble) [inline tests/unit/plugins/modules/test_s3_lifecycle_filters_are_equal.py:9]:
```suggestion


```

File: tests/unit/plugins/modules/test_s3_lifecycle_filters_are_equal.py [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+    # Test case 7: Filters with nested `And` keys differing
+    filter1 = {
+        "And": {
+            "Prefix": "test/",
+            "ObjectSizeGreaterThan": 150,
+        },
+    }
+    filter2 = {
+        "And": {
+            "Prefix": "nested/", # Different key/value
```
Comment (tremble) [inline tests/unit/plugins/modules/test_s3_lifecycle_filters_are_equal.py:81]:
```suggestion
            "Prefix": "nested/",  # Different key/value
```

File: tests/unit/plugins/modules/test_s3_lifecycle_filters_are_equal.py [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+    assert filters_are_equal(filter1, filter2) is False
+
+    # Test case 10: Filters with empty strings for `Prefix`
+    filter1 = {
+        "Prefix": ""
+    }
+    filter2 = {
+        "Prefix": ""
+    }
+    assert filters_are_equal(filter1, filter2) is True
```
Comment (tremble) [inline tests/unit/plugins/modules/test_s3_lifecycle_filters_are_equal.py:108]:
```suggestion
    assert filters_are_equal(filter1, filter2) is True
```

File: tests/unit/plugins/modules/test_s3_lifecycle_filters_are_equal.py [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+    filter1 = {
+        "And": {
+            "Prefix": "test/",
+            "ObjectSizeGreaterThan": 150,
+        },
+    }
+    filter2 = {
+        "And": {
+            "Prefix": "nested/", # Different key/value
+            "ObjectSizeLessThan": 150,
```
Comment (tremble) [inline tests/unit/plugins/modules/test_s3_lifecycle_filters_are_equal.py:82]:
```suggestion
            "ObjectSizeLessThan": 150,
```

**Review by tremble — APPROVED** [review]
Thanks,

Just as an FYI, you're still falling foul of some of our linting checks (`tox -m lint`), however they're all whitespace related so I'll push fixes.

For the unit tests, rather than quite so much copy and paste it's also possible to use pytest's `parameterize` which creates lots of copy&paste versions of the tests and means that if they start failing, it'll list all of the failed tests rather than just one.

**Pull request merged**